File: src/commonTest/kotlin/com/scoreo/infrastructure/InMemoryRepositoryTest.kt
package com.scoreo.infrastructure

import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.Match
import com.scoreo.domain.model.Player
import com.scoreo.domain.model.PlayerScore
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

class InMemoryRepositoryTest {

    @Test
    fun `PlayerRepository save deux fois avec le meme id ne cree pas de doublon`() {
        val repo = InMemoryPlayerRepository()
        val player = Player("p1", "Alice")

        repo.save(player)
        assertEquals(1, repo.getAll(includeInactive = true).size)

        repo.save(player.copy(name = "Alice modifie"))
        assertEquals(1, repo.getAll(includeInactive = true).size)
        assertEquals("Alice modifie", repo.getAll(includeInactive = true).first().name)
    }

    @Test
    fun `PlayerRepository save id different ajoute un nouvel element`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player("p1", "Alice"))
        repo.save(Player("p2", "Bob"))

        assertEquals(2, repo.getAll(includeInactive = true).size)
    }

    @Test
    fun `GameTypeRepository save deux fois avec le meme id ne cree pas de doublon`() {
        val repo = InMemoryGameTypeRepository()
        val gt = GameType("gt1", "Belote", WinCondition.HIGHEST_SCORE)

        repo.save(gt)
        assertEquals(1, repo.getAll().size)

        repo.save(gt.copy(name = "Belote modifie"))
        assertEquals(1, repo.getAll().size)
        assertEquals("Belote modifie", repo.getAll().first().name)
    }

    @Test
    fun `GameTypeRepository save id different ajoute un nouvel element`() {
        val repo = InMemoryGameTypeRepository()
        repo.save(GameType("gt1", "Belote", WinCondition.HIGHEST_SCORE))
        repo.save(GameType("gt2", "Tarot", WinCondition.LOWEST_SCORE))

        assertEquals(2, repo.getAll().size)
    }

    @Test
    fun `MatchRepository save deux fois avec le meme id ne cree pas de doublon`() {
        val repo = InMemoryMatchRepository()
        val match = Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10)), emptyList())

        repo.save(match)
        assertEquals(1, repo.getAll().size)

        repo.save(match.copy(date = 2000L))
        assertEquals(1, repo.getAll().size)
        assertEquals(2000L, repo.getAll().first().date)
    }

    @Test
    fun `MatchRepository save id different ajoute un nouvel element`() {
        val repo = InMemoryMatchRepository()
        repo.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10)), emptyList()))
        repo.save(Match("m2", 2000L, "gt1", listOf(PlayerScore("p2", 20)), emptyList()))

        assertEquals(2, repo.getAll().size)
    }

    @Test
    fun `PlayerRepository findById via save upsert conserve les donnees`() {
        val repo = InMemoryPlayerRepository()
        val player = Player("p1", "Alice")
        repo.save(player)
        val fromGetAll = repo.getAll(includeInactive = true).first()
        assertNotNull(fromGetAll)
        assertEquals("Alice", fromGetAll.name)
    }

    @Test
    fun `GameTypeRepository findById via save upsert conserve les donnees`() {
        val repo = InMemoryGameTypeRepository()
        val gt = GameType("gt1", "Belote", WinCondition.HIGHEST_SCORE)
        repo.save(gt)
        val found = repo.findById("gt1")
        assertNotNull(found)
        assertEquals("Belote", found.name)
    }

    @Test
    fun `MatchRepository findById via save upsert conserve les donnees`() {
        val repo = InMemoryMatchRepository()
        val match = Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10)), emptyList())
        repo.save(match)
        val found = repo.findById("m1")
        assertNotNull(found)
        assertEquals(1000L, found.date)
    }

    @Test
    fun `saveAll players adds all in one operation`() {
        val repo = InMemoryPlayerRepository()
        val players = listOf(
            Player(id = "p1", name = "Alice"),
            Player(id = "p2", name = "Bob"),
        )
        repo.saveAll(players)
        assertEquals(2, repo.getAll(includeInactive = true).size)
    }

    @Test
    fun `saveAll players upserts existing`() {
        val repo = InMemoryPlayerRepository()
        repo.save(Player(id = "p1", name = "Alice"))
        repo.saveAll(listOf(Player(id = "p1", name = "Updated Alice")))
        assertEquals("Updated Alice", repo.getAll(includeInactive = true).first().name)
    }
}

File: apps/scoreo/src/infrastructure/testing/inMemoryRepository.test.ts
import { describe, expect, it } from 'vitest'
import { InMemoryGameTypeRepository } from './inMemoryGameTypeRepository'
import { InMemoryMatchRepository } from './inMemoryMatchRepository'
import { InMemoryPlayerRepository } from './inMemoryPlayerRepository'

describe('InMemoryRepository fakes', () => {
  it('PlayerRepository: save twice with same id does not create a duplicate', () => {
    const repo = new InMemoryPlayerRepository()
    const player = { id: 'p1', name: 'Alice', active: true }

    repo.save(player)
    expect(repo.getAll(true)).toHaveLength(1)

    repo.save({ ...player, name: 'Alice modifie' })
    expect(repo.getAll(true)).toHaveLength(1)
    expect(repo.getAll(true)[0].name).toBe('Alice modifie')
  })

  it('PlayerRepository: save with a different id adds a new element', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    repo.save({ id: 'p2', name: 'Bob', active: true })

    expect(repo.getAll(true)).toHaveLength(2)
  })

  it('PlayerRepository: hardDelete removes the player entirely, unlike delete which soft-deletes', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: false })
    repo.save({ id: 'p2', name: 'Bob', active: true })

    repo.hardDelete('p1')

    expect(repo.getAll(true)).toEqual([{ id: 'p2', name: 'Bob', active: true }])
  })

  it('PlayerRepository: hardDelete of a nonexistent id does not throw', () => {
    const repo = new InMemoryPlayerRepository()

    expect(() => repo.hardDelete('non_existent')).not.toThrow()
  })

  it('GameTypeRepository: save twice with same id does not create a duplicate', () => {
    const repo = new InMemoryGameTypeRepository()
    const gt = {
      id: 'gt1',
      name: 'Belote',
      winCondition: 'HIGHEST_SCORE' as const,
      tieBreakRule: 'NONE' as const,
      tieBreakCondition: 'HIGHEST_SCORE' as const,
      tieBreakLabel: null,
      active: true,
    }

    repo.save(gt)
    expect(repo.getAll()).toHaveLength(1)

    repo.save({ ...gt, name: 'Belote modifie' })
    expect(repo.getAll()).toHaveLength(1)
    expect(repo.getAll()[0].name).toBe('Belote modifie')
  })

  it('GameTypeRepository: save with a different id adds a new element', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save({
      id: 'gt1',
      name: 'Belote',
      winCondition: 'HIGHEST_SCORE',
      tieBreakRule: 'NONE',
      tieBreakCondition: 'HIGHEST_SCORE',
      tieBreakLabel: null,
      active: true,
    })
    repo.save({
      id: 'gt2',
      name: 'Tarot',
      winCondition: 'LOWEST_SCORE',
      tieBreakRule: 'NONE',
      tieBreakCondition: 'HIGHEST_SCORE',
      tieBreakLabel: null,
      active: true,
    })

    expect(repo.getAll()).toHaveLength(2)
  })

  it('MatchRepository: save twice with same id does not create a duplicate', () => {
    const repo = new InMemoryMatchRepository()
    const match = {
      id: 'm1',
      date: 1000,
      gameTypeId: 'gt1',
      playerScores: [{ playerId: 'p1', score: 10 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    }

    repo.save(match)
    expect(repo.getAll()).toHaveLength(1)

    repo.save({ ...match, date: 2000 })
    expect(repo.getAll()).toHaveLength(1)
    expect(repo.getAll()[0].date).toBe(2000)
  })

  it('MatchRepository: save with a different id adds a new element', () => {
    const repo = new InMemoryMatchRepository()
    repo.save({
      id: 'm1',
      date: 1000,
      gameTypeId: 'gt1',
      playerScores: [{ playerId: 'p1', score: 10 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    })
    repo.save({
      id: 'm2',
      date: 2000,
      gameTypeId: 'gt1',
      playerScores: [{ playerId: 'p2', score: 20 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    })

    expect(repo.getAll()).toHaveLength(2)
  })

  it('PlayerRepository: findById-equivalent lookup preserves data after save upsert', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    const fromGetAll = repo.getAll(true)[0]
    expect(fromGetAll).toBeDefined()
    expect(fromGetAll.name).toBe('Alice')
  })

  it('GameTypeRepository: findById preserves data after save upsert', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save({
      id: 'gt1',
      name: 'Belote',
      winCondition: 'HIGHEST_SCORE',
      tieBreakRule: 'NONE',
      tieBreakCondition: 'HIGHEST_SCORE',
      tieBreakLabel: null,
      active: true,
    })
    const found = repo.findById('gt1')
    expect(found).toBeDefined()
    expect(found?.name).toBe('Belote')
  })

  it('MatchRepository: findById preserves data after save upsert', () => {
    const repo = new InMemoryMatchRepository()
    repo.save({
      id: 'm1',
      date: 1000,
      gameTypeId: 'gt1',
      playerScores: [{ playerId: 'p1', score: 10 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    })
    const found = repo.findById('m1')
    expect(found).toBeDefined()
    expect(found?.date).toBe(1000)
  })

  it('saveAll players adds all in one operation', () => {
    const repo = new InMemoryPlayerRepository()
    repo.saveAll([
      { id: 'p1', name: 'Alice', active: true },
      { id: 'p2', name: 'Bob', active: true },
    ])
    expect(repo.getAll(true)).toHaveLength(2)
  })

  it('saveAll players upserts existing', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    repo.saveAll([{ id: 'p1', name: 'Updated Alice', active: true }])
    expect(repo.getAll(true)[0].name).toBe('Updated Alice')
  })

  it('PlayerRepository: deleteAll clears every player', () => {
    const repo = new InMemoryPlayerRepository()
    repo.save({ id: 'p1', name: 'Alice', active: true })
    repo.save({ id: 'p2', name: 'Bob', active: false })

    repo.deleteAll()

    expect(repo.getAll(true)).toEqual([])
  })

  it('GameTypeRepository: deleteAll clears every game type', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save({
      id: 'gt1',
      name: 'Belote',
      winCondition: 'HIGHEST_SCORE',
      tieBreakRule: 'NONE',
      tieBreakCondition: 'HIGHEST_SCORE',
      tieBreakLabel: null,
      active: true,
    })

    repo.deleteAll()

    expect(repo.getAll(true)).toEqual([])
  })

  it('MatchRepository: deleteAll clears every match', () => {
    const repo = new InMemoryMatchRepository()
    repo.save({
      id: 'm1',
      date: 1000,
      gameTypeId: 'gt1',
      playerScores: [{ playerId: 'p1', score: 10 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    })

    repo.deleteAll()

    expect(repo.getAll()).toEqual([])
  })
})
